TrioDeprecationWarning: trio.MultiError is deprecated since Trio 0.22.0

For a trivial environment and *failing* test, warnings are generated due to the direct usage of multierror:
`pip install pytest-trio`
`echo "from pytest_trio.enable_trio_mode import *" > conftest.py`
`echo "async def test_warning(): assert False" > test_warning.py`
<details><summary>Output</summary>
<p>

```
(tmp) C:\Users\richa\Documents\GitHub\tmp>pytest test_warning.py
============================= test session starts =============================
platform win32 -- Python 3.10.6, pytest-7.1.3, pluggy-1.0.0
rootdir: C:\Users\richa\Documents\GitHub\tmp
plugins: trio-0.7.0
collected 1 item

test_warning.py F                                                        [100%]

================================== FAILURES =================================== 
________________________________ test_warning _________________________________ 

value = <trio.Nursery object at 0x000001A125D6B550>

    async def yield_(value=None):
>       return await _yield_(value)

..\..\..\miniconda3\envs\tmp\lib\site-packages\async_generator\_impl.py:106:    
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
..\..\..\miniconda3\envs\tmp\lib\site-packages\async_generator\_impl.py:99: in _yield_
    return (yield _wrap(value))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    async def test_warning():
>       assert False
E       assert False

test_warning.py:3: AssertionError
============================== warnings summary =============================== 
test_warning.py::test_warning
  C:\Users\richa\miniconda3\envs\tmp\lib\site-packages\pytest_trio\plugin.py:409
: TrioDeprecationWarning: trio.MultiError is deprecated since Trio 0.22.0; use B
aseExceptionGroup (on Python 3.11 and later) or exceptiongroup.BaseExceptionGroup 
(earlier versions) instead (https://github.com/python-trio/trio/issues/2211)  
    raise trio.MultiError(test_ctx.error_list)

test_warning.py::test_warning
  C:\Users\richa\miniconda3\envs\tmp\lib\site-packages\pytest_trio\plugin.py:62:
 TrioDeprecationWarning: trio.MultiError is deprecated since Trio 0.22.0; use Ba
seExceptionGroup (on Python 3.11 and later) or exceptiongroup.BaseExceptionGroup 
(earlier versions) instead (https://github.com/python-trio/trio/issues/2211)   
    if issubclass(call.excinfo.type, trio.MultiError):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info =========================== 
FAILED test_warning.py::test_warning - assert False
======================== 1 failed, 2 warnings in 0.13s ======================== 

```
</p>
</details>

--- a/newsfragments/104.feature.rst
+++ b/newsfragments/104.feature.rst
@@ -1,0 +1,10 @@
+If a test raises an ``ExceptionGroup`` (or nested ``ExceptionGroup``\ s) with only
+a single 'leaf' exception from ``pytest.xfail()`` or ``pytest.skip()``\ , we now
+unwrap it to have the desired effect on Pytest.  ``ExceptionGroup``\ s with two or
+more leaf exceptions, even of the same type, are not changed and will be treated
+as ordinary test failures.
+
+See `pytest-dev/pytest#9680 <https://github.com/pytest-dev/pytest/issues/9680>`__
+for design discussion.  This feature is particularly useful if you've enabled
+`the new strict_exception_groups=True option
+<https://trio.readthedocs.io/en/stable/reference-core.html#strict-versus-loose-exceptiongroup-semantics>`__.

--- a/newsfragments/128.misc.rst
+++ b/newsfragments/128.misc.rst
@@ -1,0 +1,4 @@
+Trio 0.22.0 deprecated ``MultiError`` in favor of the standard-library
+(or `backported <https://pypi.org/project/exceptiongroup/>`__) ``ExceptionGroup``
+type; ``pytest-trio`` now uses ``ExceptionGroup`` and therefore requires
+Trio 0.22.0 or later.

--- a/pytest.ini
+++ b/pytest.ini
@@ -1,2 +1,7 @@
 [pytest]
 addopts = -ra -v --pyargs pytest_trio --verbose --cov
+filterwarnings =
+    error
+    default::pytest.PytestAssertRewriteWarning
+    default::pytest.PytestDeprecationWarning
+    default::pytest.PytestUnraisableExceptionWarning

--- a/pytest_trio/_tests/test_async_yield_fixture.py
+++ b/pytest_trio/_tests/test_async_yield_fixture.py
@@ -300,4 +300,6 @@ def test_async_yield_fixture_crashed_teardown_allow_other_teardowns(testdir):
     result = testdir.runpytest()
 
     result.assert_outcomes(failed=1, passed=2)
-    result.stdout.re_match_lines([r"E\W+RuntimeError: Crash during fixture teardown"])
+    result.stdout.re_match_lines(
+        [r"(E\W+| +\| )RuntimeError: Crash during fixture teardown"]
+    )

--- a/pytest_trio/_tests/test_basic.py
+++ b/pytest_trio/_tests/test_basic.py
@@ -74,3 +74,52 @@ def test_sync_function_with_trio_mark(testdir):
     result = testdir.runpytest()
 
     result.assert_outcomes(errors=1)
+
+
+def test_skip_and_xfail(testdir):
+
+    testdir.makepyfile(
+        """
+        import functools
+        import pytest
+        import trio
+
+        trio.run = functools.partial(trio.run, strict_exception_groups=True)
+
+        @pytest.mark.trio
+        async def test_xfail():
+            pytest.xfail()
+
+        @pytest.mark.trio
+        async def test_skip():
+            pytest.skip()
+
+        async def callback(fn):
+            fn()
+
+        async def fail():
+            raise RuntimeError
+
+        @pytest.mark.trio
+        async def test_xfail_and_fail():
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(callback, pytest.xfail)
+                nursery.start_soon(fail)
+
+        @pytest.mark.trio
+        async def test_skip_and_fail():
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(callback, pytest.skip)
+                nursery.start_soon(fail)
+
+        @pytest.mark.trio
+        async def test_xfail_and_skip():
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(callback, pytest.skip)
+                nursery.start_soon(callback, pytest.xfail)
+    """
+    )
+
+    result = testdir.runpytest("-s")
+
+    result.assert_outcomes(skipped=1, xfailed=1, failed=3)

--- a/pytest_trio/_tests/test_sync_fixture.py
+++ b/pytest_trio/_tests/test_sync_fixture.py
@@ -139,4 +139,6 @@ def test_sync_yield_fixture_crashed_teardown_allow_other_teardowns(testdir):
     result = testdir.runpytest()
 
     result.assert_outcomes(failed=1, passed=2)
-    result.stdout.re_match_lines([r"E\W+RuntimeError: Crash during fixture teardown"])
+    result.stdout.re_match_lines(
+        [r"(E\W+| +\| )RuntimeError: Crash during fixture teardown"]
+    )

--- a/pytest_trio/plugin.py
+++ b/pytest_trio/plugin.py
@@ -10,6 +10,7 @@ import pytest
 import trio
 from trio.abc import Clock, Instrument
 from trio.testing import MockClock
+from _pytest.outcomes import Skipped, XFailed
 
 if sys.version_info[:2] < (3, 11):
     from exceptiongroup import BaseExceptionGroup
@@ -343,7 +344,25 @@ def _trio_test(run):
                     f"Expected at most one Clock in kwargs, got {clocks!r}"
                 )
             instruments = [i for i in kwargs.values() if isinstance(i, Instrument)]
-            return run(partial(fn, **kwargs), clock=clock, instruments=instruments)
+            try:
+                return run(partial(fn, **kwargs), clock=clock, instruments=instruments)
+            except BaseExceptionGroup as eg:
+                queue = [eg]
+                leaves = []
+                while queue:
+                    ex = queue.pop()
+                    if isinstance(ex, BaseExceptionGroup):
+                        queue.extend(ex.exceptions)
+                    else:
+                        leaves.append(ex)
+                if len(leaves) == 1:
+                    if isinstance(leaves[0], XFailed):
+                        pytest.xfail()
+                    if isinstance(leaves[0], Skipped):
+                        pytest.skip()
+                # Since our leaf exceptions don't consist of exactly one 'magic'
+                # skipped or xfailed exception, re-raise the whole group.
+                raise
 
         return wrapper
 

--- a/pytest_trio/plugin.py
+++ b/pytest_trio/plugin.py
@@ -1,6 +1,6 @@
 """pytest-trio implementation."""
+import sys
 from functools import wraps, partial
-from traceback import format_exception
 from collections.abc import Coroutine, Generator
 from contextlib import asynccontextmanager
 from inspect import isasyncgen, isasyncgenfunction, iscoroutinefunction
@@ -10,6 +10,9 @@ import pytest
 import trio
 from trio.abc import Clock, Instrument
 from trio.testing import MockClock
+
+if sys.version_info[:2] < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
 
 ################################################################
 # Basic setup
@@ -50,13 +53,6 @@ def pytest_configure(config):
         "markers",
         "trio: mark the test as an async trio test; it will be run using trio.run",
     )
-
-
-@pytest.hookimpl(tryfirst=True)
-def pytest_exception_interact(node, call, report):
-    if issubclass(call.excinfo.type, trio.MultiError):
-        # TODO: not really elegant (pytest cannot output color with this hack)
-        report.longrepr = "".join(format_exception(*call.excinfo._excinfo))
 
 
 ################################################################
@@ -407,8 +403,12 @@ def _trio_test_runner_factory(item, testfunc=None):
                     )
                 )
 
-        if test_ctx.error_list:
-            raise trio.MultiError(test_ctx.error_list)
+        if len(test_ctx.error_list) == 1:
+            raise test_ctx.error_list[0]
+        elif test_ctx.error_list:
+            raise BaseExceptionGroup(
+                "errors in async test and trio fixtures", test_ctx.error_list
+            )
 
     _bootstrap_fixtures_and_run_test._trio_test_runner_wrapped = True
     return _bootstrap_fixtures_and_run_test

